bluesky/plans.py

```diff
@@ -2691,6 +2691,148 @@ def inner_relative_adaptive_scan():
+    if (num_points - 1) <= 2*step_factor:
+        raise ValueError(
+            "Increase num_points and/or decrease step_factor"
+            " or tune_centroid will never converge to a solution"
+        )
+    _md = {'detectors': [det.name for det in detectors],
+           'motors': [motor.name],
+           'plan_args': {'detectors': list(map(repr, detectors)),
+                         'motor': repr(motor),
+                         'start': start,
+                         'stop': stop,
+                         'num_points': num_points,
+                         'min_step': min_step,},
+           'plan_name': 'tune_centroid',
+           'hints': {},
+          }
+    _md.update(md or {})
+    try:
+        dimensions = [(motor.hints['fields'], 'primary')]
+    except (AttributeError, KeyError):
+        pass
+    else:
+        _md['hints'].setdefault('dimensions', dimensions)
+
+    @stage_decorator(list(detectors) + [motor])
+    @run_decorator(md=_md)
+    def _tune_core(start, stop, num_points, signal):
+        next_pos = start
+        step = (stop - start) / (num_points - 1)
+        peak_position = None
+        cur_I = None
+        cur_det = {}
+        sum_I = 0       # for peak centroid calculation, I(x)
+        sum_xI = 0
+        
+        while abs(step) >= min_step:
+            yield Msg('checkpoint')
+            yield from mv(motor, next_pos)
+            yield Msg('create', None, name='primary')
+            for det in detectors:
+                yield Msg('trigger', det, group='B')
+            yield Msg('wait', None, 'B')
+            for det in separate_devices(detectors + [motor]):
+                cur_det = yield Msg('read', det)
+                if signal in cur_det:
+                    cur_I = cur_det[signal]['value']
+                    sum_I += cur_I
+                    position = motor.read()[motor.name]["value"]
+                    sum_xI += position * cur_I
+
+            yield Msg('save')
+
+            if (stop - start) < abs(stop - start):
+                in_range = start >= next_pos >= stop  # negative motion
+            else:
+                in_range = start <= next_pos <= stop  # positive motion
+            
+            if in_range:
+                next_pos += step
+            else:
+                if sum_I == 0:
+                    return
+                peak_position = sum_xI / sum_I  # centroid
+                # TODO: report current peak_position in metadata
+                # RE.md["peak_position"] = str(peak_position)
+                start = peak_position - step_factor*step
+                stop = peak_position + step_factor*step
+                if snake:
+                    start, stop = stop, start
+                step = (stop - start) / (num_points - 1)
+                next_pos = start
+
+        # finally, move to peak position
+        if peak_position is not None:
+            # RE.md["peak_position"] = peak_position
```

Comment:
This is too late to make it into _this_ runs meta-data, but it will appear in the _next_ runs which would be extremely confusing.

Reply:
Agreed.  Will mark this as a possible improvement.  And note in the docstring how to retrieve the result from the motor position (in keeping with the philosophy that plan results are retrieved from hardware).